### PR TITLE
[LBSE] Add RenderSVGResourceMasker

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 08ff1e4f09f8c06c8ffae86dd8e1bb4feed8abe5
<pre>
[LBSE] Add RenderSVGResourceMasker
<a href="https://bugs.webkit.org/show_bug.cgi?id=264049">https://bugs.webkit.org/show_bug.cgi?id=264049</a>

Reviewed by Nikolas Zimmermann.

Implement the second SVG resource in LBSE: masking (RenderSVGResourceMasker).
Mask content invalidations/changes are supported.

The resource takes cyclic references into account.

Covered by existing tests (dozens of progressions, see TestExpectations).

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-intro-01-f-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-mask-01-b-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-root-position-masking-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/empty-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-mask-and-object-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-mask-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-changes-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-child-changes-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-colorspace-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-excessive-malloc-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-inside-defs-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-invalidation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-with-all-units-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-with-default-value-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/transformedMaskFails-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/visibility-override-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/filters/filter-clip-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/foreignObject/mask-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/text-mask-update-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink.html: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-shrink.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgMaskerResourceFromStyle const):
(WebCore::RenderLayerModelObject::paintSVGClippingMask const):
(WebCore::RenderLayerModelObject::paintSVGMask const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourceContainer const):
(WebCore::RenderObject::isRenderSVGResourceMasker const):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp: Added.
(WebCore::RenderSVGResourceMasker::RenderSVGResourceMasker):
(WebCore::createImageBuffer):
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceMaskerInlines.h: Added.
(WebCore::RenderSVGResourceMasker::maskElement const):
(WebCore::RenderSVGResourceMasker::maskUnits const):
(WebCore::RenderSVGResourceMasker::maskContentUnits const):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::paintSVGClippingMask): Deleted.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::createElementRenderer):
(WebCore::SVGMaskElement::calculateMaskContentRepaintRect):
* Source/WebCore/svg/SVGMaskElement.h:

Canonical link: <a href="https://commits.webkit.org/271153@main">https://commits.webkit.org/271153@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e388f9d3bf5bca80e4597a778047f6ce0e819735

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27502 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6144 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29730 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25177 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3539 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24966 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4930 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23613 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4293 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24610 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30366 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25035 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30591 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2615 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24345 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6617 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4865 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/svg/RenderSVGContainer.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGContainer.cpp]: I think it's time to move them into a shared place, that is not SVGRenderSupport, but REnderLayerModelObject.h. There is also a section in that header, which already contains LBSE specific functions shared across RenderSVG* renderers. Let's move paintSVGClipMask and paintSVGMask there :-)

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Soon this is obsolete. We can land this before the resource invalidation rework.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: auto?

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Not sure how relevant that is in LBSE.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: auto?

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: auto?

File: Source/WebCore/rendering/svg/SVGRenderSupport.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGRenderSupport.cpp]: Evil boiler plate stuff, will get better with my patch to rework the SVG resource lookup.

File: Source/WebCore/rendering/svg/SVGRenderingContext.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGRenderingContext.cpp]: Do we really want to add stuff to SVGRenderingContext? I am trying to make SVGRenderSupport / SVGRenderingContext legacy-only.

**Review by nikolaszimmermann — COMMENTED** [review]
Great first stab.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Good point ,should it be an ASSERT instead or just remove the check?

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Fixed.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Fixed.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Sure, fixed.

File: Source/WebCore/rendering/svg/RenderSVGContainer.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGContainer.cpp]: You are right, that way we can also make it non-static. Done.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: I'd say try the assert. Not sure if that's already possible before the LBSE rework patch landed (regarding invalidation).

File: Source/WebCore/rendering/svg/SVGRenderingContext.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/SVGRenderingContext.cpp]: Sure, removed.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Sure, done!

File: Source/WebCore/rendering/RenderElement.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/RenderElement.cpp]: Really needed?

File: Source/WebCore/rendering/RenderLayerModelObject.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/RenderLayerModelObject.cpp:590]: The whole method needs a `#if ENABLE(LAYER_BASED_SVG_ENGINE)' guard.

File: Source/WebCore/rendering/RenderLayerModelObject.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/RenderLayerModelObject.cpp]: IIRC we only call this for `PaintPhase::Mask`, no? Then you can add the same assertion as in `paintSVGClippingMask` just for `PaintPhase::Mask`.

File: Source/WebCore/rendering/RenderLayerModelObject.h
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/RenderLayerModelObject.h]: The necessity of 'adjustedPaintOffset' (the 2nd parameter) or the absence of it in `paintSVGClippingMask` should be revisited / clarified. IMHO one should spell out the parameter name here for clarity - if it doesn't violate style rules.

File: Source/WebCore/rendering/RenderObject.h
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/RenderObject.h]: We should decide on a one, unique solution. Clipper has LegacySVGResourceClipper / SVGResourceClipper. For marker we only have SVGResourceMarker at present. I'd go for explicit LegacySVGResourceMasker / SVGResourceMasker.

File: Source/WebCore/rendering/svg/RenderSVGContainer.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGContainer.cpp:130]: Much better :-)

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: We still have DIRECT2D support?

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: s/GC/auto/

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:69]: IIRC we have a generic function that can be used to create image buffers. Maybe I'm wrong. Let's check :-)

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: I think this should go - and be in sync with how clipper does things.

File: Source/WebCore/rendering/svg/SVGResources.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGResources.cpp:235]: This is no longer needed, LBSE doesn't use SVGResources anymore.

File: Source/WebCore/svg/SVGMaskElement.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/svg/SVGMaskElement.cpp]: auto in a few places.

Comment (rwlbuis) [reply Source/WebCore/svg/SVGMaskElement.cpp]: Fixed.

File: Source/WebCore/rendering/svg/SVGResources.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/SVGResources.cpp:235]: Good point, removed.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Done.

File: Source/WebCore/rendering/RenderLayerModelObject.h
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderLayerModelObject.h]: I added the name, indeed it can't be deduced from the parameter type.

File: Source/WebCore/rendering/RenderLayerModelObject.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderLayerModelObject.cpp]: Added the ASSERT.

File: Source/WebCore/rendering/RenderLayerModelObject.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderLayerModelObject.cpp:590]: It actually is already included in one.

File: Source/WebCore/rendering/RenderElement.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderElement.cpp]: Looks like not needed, removed.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Indeed that assertion is not relevant anymore, removed.

File: Source/WebCore/rendering/RenderObject.h
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderObject.h]: You are right, confusing to be inconsistent, I added LegacySVGResourceMasker.

File: Source/WebCore/rendering/RenderLayerModelObject.cpp
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/RenderLayerModelObject.cpp]: The ::Mask condition in the if can go away.

**Review by nikolaszimmermann — COMMENTED** [review]
Looks great already.

File: Source/WebCore/rendering/RenderLayerModelObject.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderLayerModelObject.cpp]: Wel spotted, removed.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:69]: I tried quickly and saw a regression in svg/custom (did not try other directories). Since this could be some subtle code difference we agreed to leave this for later.

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: No, seems removed, fixed.

**Review by nikolaszimmermann — APPROVED** [review]
Allright, let's continue when this is in.

**Pull request merged**
